File: app/src/main/kotlin/com/omerilhanli/appcentchallenge/asistive/tool/AutoFitGridLayoutManager.java
package com.omerilhanli.appcentchallenge.asistive.tool;

import android.content.Context;
import android.content.res.Resources;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import static android.content.res.Configuration.ORIENTATION_LANDSCAPE;

public class AutoFitGridLayoutManager extends GridLayoutManager {

    private final static int COUNT = 2;

    private int columnWidth;
    private boolean columnWidthChanged = true;
    private Context mContext;

    public AutoFitGridLayoutManager(Context context) {
        super(context, COUNT);
        this.mContext = context;
        setColumnWidth(getScreenWidth() / COUNT);
    }

    private int getScreenWidth() {
        return Resources.getSystem().getDisplayMetrics().widthPixels;
    }

    private void setColumnWidth(int newColumnWidth) {
        if (newColumnWidth > 0 && newColumnWidth != columnWidth) {
            columnWidth = newColumnWidth;
            columnWidthChanged = true;
        }
    }

    @Override
    public void onLayoutChildren(RecyclerView.Recycler recycler, RecyclerView.State state) {
        if (columnWidthChanged && columnWidth > 0) {
            int totalSpace = getWidth() - getPaddingRight() - getPaddingLeft();
            int orientation = mContext.getResources().getConfiguration().orientation;
            int spanCount = Math.max(COUNT, totalSpace / columnWidth);
            if (orientation == ORIENTATION_LANDSCAPE) {
                spanCount++;
            }
            setSpanCount(spanCount);
            columnWidthChanged = false;
        }
        super.onLayoutChildren(recycler, state);
    }
}

File: app/src/main/kotlin/com/omerilhanli/appcentchallenge/asistive/tool/AutoFitGridLayoutManager.kt
package com.omerilhanli.appcentchallenge.asistive.tool

import android.content.Context
import android.content.res.Configuration
import android.content.res.Resources
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.Recycler
import kotlin.math.max

/**
 * Recyclerview için listedekileri grid yapısında otomatik olarak fitleyen custom manager class'ı.
 */
class AutoFitGridLayoutManager(private val mContext: Context) :
    GridLayoutManager(mContext, COUNT) {
    private var columnWidth = 0
    private var columnWidthChanged = true
    private val screenWidth: Int
        get() = Resources.getSystem().displayMetrics.widthPixels

    private fun setColumnWidth(newColumnWidth: Int) {
        if (newColumnWidth > 0 && newColumnWidth != columnWidth) {
            columnWidth = newColumnWidth
            columnWidthChanged = true
        }
    }

    override fun onLayoutChildren(
        recycler: Recycler,
        state: RecyclerView.State
    ) {
        if (columnWidthChanged && columnWidth > 0) {
            val totalSpace = width - paddingRight - paddingLeft
            val orientation = mContext.resources.configuration.orientation
            var spanCount = max(
                COUNT,
                totalSpace / columnWidth
            )
            if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
                spanCount++
            }
            setSpanCount(spanCount)
            columnWidthChanged = false
        }
        super.onLayoutChildren(recycler, state)
    }

    companion object {
        private const val COUNT = 2
    }

    init {
        setColumnWidth(screenWidth / COUNT)
    }
}